Add UR estimation to the osu!mania ruleset

Natelytle opened the pull request:
Massive thanks to Frost for the majority of the math behind this rework, and to Evening, Molneya, and Komirin for guidance.
# Estimates Unstable Rate and replaces Accuracy with it
## Changes
- Raw accuracy percentages are no longer used to scale performance
- Instead, the UR value of the score is **estimated from the judgements of the play** and the sizes of the hit windows, and used to scale performance in an **hit window agnostic** way
## Reasoning
- Between Scorev1, Scorev2, and osu!lazer Score, the way hit windows work differ massively. Because of this, the same play will receive a different accuracy value depending on which system is used.
- Performance previously did not scale with OD.
-  Unstable rate is an easier metric to work with, as there is a true "perfect" value.
## Estimation Theory
In order to estimate UR, we assume all hits are normally distributed, with a mean of ±0 and a deviation σ. This gives us the probability that with a certain σ, any given hit lands in a given window. We can compare these percentages to the true percentages of any given judgement in a play, and return whichever σ value is the closest match.
![image](https://github.com/ppy/osu/assets/92956514/fc878f88-fd64-4b33-a54f-a300f842ae6b)
Further documentation can be found in a google doc [here.](https://docs.google.com/document/d/1IC2oFnQn1shhin6NdyZHwM0nUmA3c2ZVqpK8pppdmFU/edit)


## Considerations
- Mania hit windows may potentially change as mania in lazer is ironed out. This will require addressing and an update to the current tests.
-  Classic mod doesn't change the results of a replay, forcing me to use a workaround that causes note-only classic mod plays to give incorrect values when replayed.
- Requires MathNet.Numerics, a package for advanced mathematical formulas not present in C#.

---

SR/PP spreadsheet: https://docs.google.com/spreadsheets/d/1SBFrFOWIxJM-gACZrOk-2I6mH35MN9FRUDXbrTcGnJs/edit
As of a2197e2

Natelytle committed:
add deviation estimation to mania's perfcalc

Natelytle committed:
add UR attribute, change initial min to 300UR

Natelytle committed:
bound deviation estimate to numbers greater than 0

Natelytle committed:
fix hitwindows

Natelytle committed:
account for converts, add note counts

Natelytle committed:
use onlineid for contingency

Natelytle committed:
fix convert hit windows

Natelytle committed:
account for HR/EZ in deviation calculation

Natelytle committed:
use erfc approximation

Natelytle committed:
Add rudimentary difficulty scaling curve

Natelytle committed:
Change curve

Natelytle committed:
Don't give PP to jittered plays

Natelytle committed:
merge master and change hit window formulas

Natelytle committed:
change hit window formulas

Natelytle committed:
fix judgement

Natelytle committed:
lazer deviation is finally accurate woo

Natelytle committed:
increase totalhits by 1

Natelytle committed:
quick legacy LN idea

Natelytle committed:
Fix formatting

Natelytle committed:
change legacy LN calculation

Natelytle committed:
switch to totalvalue

Natelytle committed:
add comments

Natelytle committed:
rewrite everything with erf for simplicity

Natelytle committed:
change isLegacy toggle

Natelytle committed:
whop

Natelytle committed:
readability bump

Natelytle committed:
rename method

Natelytle committed:
Merge remote-tracking branch 'master/master' into maniastatacc

Natelytle committed:
fix convert OD with lazer judgements

Natelytle committed:
Revert unrelated files changed in formatting commit

Eve-ning commented on osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs:
```diff
@@ -55,30 +58,212 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
         }
 
         private double computeDifficultyValue(ManiaDifficultyAttributes attributes)
         {
-            double difficultyValue = Math.Pow(Math.Max(attributes.StarRating - 0.15, 0.05), 2.2) // Star rating to pp curve
-                                     * Math.Max(0, 5 * scoreAccuracy - 4) // From 80% accuracy, 1/20th of total pp is awarded per additional 1% accuracy
-                                     * (1 + 0.1 * Math.Min(1, totalHits / 1500)); // Length bonus, capped at 1500 notes
+            double difficultyValue = Math.Pow(Math.Max(attributes.StarRating - 0.15, 0.05), 2.2); // Star rating to pp curve
+
+            difficultyValue *= Math.Max(1.2 * Math.Pow(SpecialFunctions.Erf(300 / estimatedUR), 1.6) - 0.2, 0);
```
Please help clarify what is this change with graphs

Natelytle replied:
added a desmos link

Eve-ning commented on osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs:
```diff
@@ -55,30 +58,212 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+        {
+            double[] judgements = new double[5];
+
+            var track = new TrackVirtual(10000);
+            score.Mods.OfType<IApplicableToTrack>().ForEach(m => m.ApplyToTrack(track));
+            double clockRate = track.Rate;
+
+            double windowMultiplier = 1 / clockRate;
+
+            if (score.Mods.Any(m => m is ModHardRock))
+                windowMultiplier *= 1 / 1.4;
+            else if (score.Mods.Any(m => m is ModEasy))
+                windowMultiplier *= 1.4;
+
+            if (attributes.OverallDifficulty < 5)
+                judgements[0] = (22.4 - 0.6 * attributes.OverallDifficulty) * windowMultiplier;
+            else
+                judgements[0] = (24.9 - 1.1 * attributes.OverallDifficulty) * windowMultiplier;
+            judgements[1] = (64 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            judgements[2] = (97 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            judgements[3] = (127 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            judgements[4] = (151 - 3 * attributes.OverallDifficulty) * windowMultiplier;
```
I don't get how the clockrate is related to the judgements

Natelytle replied:
In lazer, the judgement windows for all 4 gamemodes are multiplied by the rate, as outlined in https://github.com/ppy/osu-wiki/pull/8791

Eve-ning reviewed (commented):
I recommend documenting this somewhere accessible, then commenting it onto the code.
The math is rather deep, thus it'll pose difficulties for future programmers that take over this code, or wants to overwrite it.

Natelytle committed:
Add desmos links to likelihood functions

Natelytle committed:
Code quality overhaul

Natelytle committed:
fix const name

Natelytle committed:
More formatting

Natelytle committed:
Remove pointless erf approximation, fix lazer return statement

Natelytle committed:
Massively improve estimation accuracy using a folded distribution

Natelytle committed:
update desmos, remove redundant parenthesis

Natelytle committed:
Reformatting, increase precision a lot by rewriting in terms of the log of the complement

Natelytle committed:
Rename "Judgements" to "HitWindows"

Eve-ning commented on osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs:
```diff
@@ -55,30 +59,219 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+        {
+            double[] hitWindows = new double[5];
+
+            var track = new TrackVirtual(10000);
+            score.Mods.OfType<IApplicableToTrack>().ForEach(m => m.ApplyToTrack(track));
+            double clockRate = track.Rate;
```
I got your comment on a prior thread, I recommend to comment this code as it's quite vague. 
I'm uncertain if it's a standard/common way to yield the clock rate

Natelytle replied:
Added comments.

Eve-ning commented on osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs:
```diff
@@ -55,30 +59,219 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+            double maxVal = Math.Max(l1, l2);
+            double minVal = Math.Min(l1, l2);
+
+            if (double.IsNegativeInfinity(maxVal))
```
1) Leave a short comment on when a negative infinity occurs
2) Are callers of this function handling the return value `-inf` correctly?

Natelytle replied:
This review made me realize a possible NaN in the logDiff function, but after fixing it everything handles -inf perfectly fine.
![image](https://user-images.githubusercontent.com/92956514/219976315-ad615d63-3e6a-454f-a954-a2f853828607.png)

Eve-ning reviewed (commented):
This is my 2nd review, please take a look at it.

Generally, I think the code is acceptable. 
However, from a laypeople perspective, it's definitely rough to understand the theory. One way you can bridge the knowledge gap is to create some unit tests showing that what you're giving, is what they expect.

For example, if your pull request estimates UR well enough, then
```py
assert abs(computeEstimatedUr(...) - computeTrueUr(...)) < epsilon
```
This way, maintainers can be assured that your code is numerically stable and correct, even on edge cases.

I understand that your model assumes some traits of a real score, thus can't be generated like so:
```py
errors = normal_rand(n=n_samples)
judgements = categorize_errors(errors)
assert abs(computeEstimatedUr(judgements) - computeTrueUr(errors)) < epsilon
```

Thus, it'd be good to have some raw data that can be ran through a unit test.

Natelytle commented on osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs:
```diff
@@ -55,30 +59,219 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+                double p50Tail = logDiff(logPcNote(h100 * 1.5, d), logPcNote(h50 * 1.5, d));
+                double p0Tail = logPcNote(h50 * 1.5, d);
+
+                double pMax = logSum(pMaxNote + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), pMaxTail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p300 = logSum(p300Note + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), p300Tail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p200 = logSum(p200Note + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), p200Tail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p100 = logSum(p100Note + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), p100Tail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p50 = logSum(p50Note + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), p50Tail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p0 = logSum(p0Note + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), p0Tail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+
+                double gradient = Math.Exp(
+                    (countPerfect * pMax
+                     + (countGreat + 0.5) * p300
+                     + countGood * p200
+                     + countOk * p100
+                     + countMeh * p50
+                     + countMiss * p0) / totalJudgements
+                );
+
+                return -gradient;
```
Done.

Natelytle replied:
I've tried writing it as a loop over an array, but it looks really ugly in my opinion and isn't much more concise.

Natelytle commented on osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs:
```diff
@@ -55,30 +59,219 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+
+                double p0Hold = combinedProb(logPc50Head, logPc50Tail);
+
+                double pMax = logSum(pMaxNote + Math.Log(attributes.NoteCount), pMaxHold + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p300 = logSum(p300Note + Math.Log(attributes.NoteCount), p300Hold + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p200 = logSum(p200Note + Math.Log(attributes.NoteCount), p200Hold + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p100 = logSum(p100Note + Math.Log(attributes.NoteCount), p100Hold + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p50 = logSum(p50Note + Math.Log(attributes.NoteCount), p50Hold + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p0 = logSum(p0Note + Math.Log(attributes.NoteCount), p0Hold + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
```
Done.

Natelytle commented on osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs:
```diff
@@ -38,7 +42,7 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
             countMeh = score.Statistics.GetValueOrDefault(HitResult.Meh);
             countMiss = score.Statistics.GetValueOrDefault(HitResult.Miss);
-            scoreAccuracy = calculateCustomAccuracy();
+            estimatedUr = computeEstimatedUr(score, maniaAttributes) * 10;
```
Done.

Natelytle committed:
Address reviews

Natelytle committed:
account for mania tests

Natelytle committed:
Fix issue with high misscount plays returning 115 UR

Natelytle committed:
Fix incorrect logDiff function, make lazer tail multiplier a const

abraker95 commented on osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs:
```diff
@@ -55,30 +61,243 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+        // Log rules make addition and subtraction of the non-log value non-trivial, these methods simply add and subtract the base value of logs.
+        private double logSum(double l1, double l2)
+        {
+            double maxVal = Math.Max(l1, l2);
+            double minVal = Math.Min(l1, l2);
```
Maybe a nitpick, but variables `l1` and `l2` look like numerical values, which doesn't help with code readability. Same in `logDiff` below.

Natelytle replied:
Changed to `firstLog` and `secondLog`

Natelytle committed:
Address abraker's review, add approximation for legacy LN tails

Natelytle committed:
Address abraker's review, add approximation for legacy LN tails

Natelytle committed:
Merge remote-tracking branch 'origin/maniastatacc' into maniastatacc

Natelytle commented:
sorry, I'm not sure what happened with those commits

Natelytle committed:
Fix NaNs (whoops)

Natelytle committed:
Avoid repeated code using structs and methods

Natelytle committed:
Add comments

Natelytle committed:
Change types, remove pointless duplicate likelihood function

Natelytle committed:
Fix erfc approximation

Eve-ning reviewed (changes requested):
Code wise, looks fine. 

However, I cannot, in confidence, explicitly approve this for the whole team as I'm not a expert in this field.

There's 2 ways that you can move forward with this PR
1) Find someone who's can truly understand the whole PR to approve it.
2) Create unit tests that show that what you've done is what the team expects it to do.

I highly recommend the latter, as it's good coding practice and makes the code more durable.
Consider the scenario that somehow, a PR was made outside of these files that changes the hit windows. 

E.g. These magic constants were changed
```cs
if (attributes.OverallDifficulty < 5)
    hitWindows[0] = (22.4 - 0.6 * attributes.OverallDifficulty) * windowMultiplier;
else
    hitWindows[0] = (24.9 - 1.1 * attributes.OverallDifficulty) * windowMultiplier;
hitWindows[1] = (64 - 3 * attributes.OverallDifficulty) * windowMultiplier;
hitWindows[2] = (97 - 3 * attributes.OverallDifficulty) * windowMultiplier;
hitWindows[3] = (127 - 3 * attributes.OverallDifficulty) * windowMultiplier;
hitWindows[4] = (151 - 3 * attributes.OverallDifficulty) * windowMultiplier;
```

This will essentially go undetected if there's no tests safeguarding its behavior. Trying to backtrace the root cause will take lots of effort.

By creating unit tests, we, as a team, can be sure that this change is as expected, and will continue be as expected for the coming months. Furthermore, we'll (including you) get notified (through test failures) if the code somehow breaks on edge cases.

Natelytle committed:
expose hit windows publicly for testing purposes

Natelytle committed:
Add tests

Natelytle commented:
I can't see a clear way to test for the multiplier on lazer LN tails, or on whether or not rate adjustments affect the mania hit windows, but I implemented some simple tests to see if changes affect the UR estimation or if the hit window formulas are changed. I'm considering adding a test to compare the estimated UR of a play to the real play, but that may require more work than is worth it.

Natelytle committed:
Edit comments

Eve-ning commented on osu.Game.Rulesets.Mania.Tests/ManiaUnstableRateEstimationTest.cs:
```diff
@@ -0,0 +1,114 @@
+            }, attributes);
+
+            // Platform-dependent math functions (Pow, Cbrt, Exp, etc) and advanced math functions (Erf, FindMinimum) may result in slight differences.
+            Assert.That(perfAttributes.EstimatedUr, Is.EqualTo(expectedEstimatedUnstableRate).Within(0.001));
```
Add a assert message on failure as the 3rd argument:
`That(Object, IResolveConstraint, String)`

1) We can see what failed in english
2) How badly it failed

`That(..., ..., "Estimated UR differed by ...")`

https://learn.microsoft.com/en-us/dotnet/api/nunit.framework.assert.that?view=xamarin-ios-sdk-12#nunit-framework-assert-that(system-object-nunit-framework-constraints-iresolveconstraint-system-string-system-object())

Natelytle replied:
Done.

Eve-ning reviewed (changes requested):
looks much better, you've made it easier for the laymen to verify the behvaior of the code

A few additional comments

1) In the PR description state the motivation of the PR, not just what the PR is doing. i.e. you've estimated the UR of a play through judgements, so what comes next from this? what's the follow up purpose?
After a quick chat, maybe include that your implementation allows UR to be calculated without simulation.
2) Add a brief comment/docstring for each test. I.e. `// Test that our ur estimation is within +-x`.

I think this is fine as is, it's likely this will be revisited when another component such as pp SR builds upon this.

Natelytle committed:
Add comments and fail messages to UR estimation tests

Natelytle committed:
Change UR scaling curve

Natelytle committed:
Revert to live scaling

Eve-ning reviewed (dismissed):
lgtm

abraker95 reviewed (dismissed):
![image](https://user-images.githubusercontent.com/4358724/232323562-860c2ebc-39b8-4ebf-824e-210832290301.png)

apollo-dw commented on osu.Game.Rulesets.Mania/Difficulty/ManiaDifficultyAttributes.cs:
```diff
@@ -19,6 +19,27 @@ public class ManiaDifficultyAttributes : DifficultyAttributes
+        /// <summary>
+        /// Conversion status from standard.
+        /// </summary>
+        public bool Convert { get; set; }
```
nit: change to `IsConvert`, should yield better readibility when doing `if (attributes.IsConvert)`

Natelytle replied:
Done.

apollo-dw commented on osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs:
```diff
@@ -55,30 +72,249 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+            double[] legacyHitWindows = new double[5];
+
+            double overallDifficulty = attributes.OverallDifficulty;
+
+            if (attributes.Convert)
+                overallDifficulty = 10;
```
according to the information that will be added in https://github.com/ppy/osu-wiki/pull/8245, this can actually be either OD 10 or OD ~5 (see [wiki/Gameplay/Judgement/osu!mania/en.md](https://github.com/ppy/osu-wiki/commit/3795d6d161a18dde46cc2c1a30dd6835e71b1a0f#diff-182fdf1bbbc79c215c413a619f2aacd28cb4f82e32cfa813c5e3d0eb7d101171)). this should definitely be accounted for.

aside from that, this is definitely going to need a comment. i've discussed this with natelytle in pp dev discord and here's a potential comment i can propose:

```
When converting beatmaps to osu!mania in stable, the resulting hit window sizes are dependent on whether the beatmap's OD is above or below 4.
```

Natelytle replied:
Done.

Natelytle committed:
Rename to IsConvert, add handling for converts below od4

Natelytle committed:
Move comment

Natelytle committed:
Revert back to previous curve, add more tests

Natelytle commented:
Decided to go back to adding the new curve in this PR after a discussion with Evening, @smoogipoo if you find some time could you make a smoogisheet for this? Be sure to append classic to the scores before feeding them into the sheet, as this rework returns lazer values for non classic mod scores.

Natelytle committed:
Update desmos

Natelytle committed:
Change curve to clamp PP to <700ur

Natelytle committed:
Change hit window test to not require a beatmap, test multiple ODs

Natelytle committed:
Change curve (again) to fix jittering edge cases

Natelytle committed:
Merge remote-tracking branch 'osumaster/master' into maniastatacc

Eve-ning reviewed (approved):
Hello,

I'm approving these changes on behalf of the commitee
1) Though the theory behind it is complex, it's well tested with test-cases. 
2) Furthermore, we've conducted player feedback on this PR for a few months now, and have no complaints lately. Thus it's unlikely we've missed an edge case

---

I'd personally prefer the PR description to be more digestable like:

# Adds Accuracy to UR Estimation

## Changes
pp calculation depends on the estimated UR from judgements instead of accuracy

## Why
Accuracy is a less preferably metric
- Accuracy is less fine-grained than UR due to data-binning to judgements.
- The definition of accuracy changes between score v1 and v2.

UR is independent of the way judgements are binned. 
However, it's not possible to find the true UR exactly
Thus, we estimate UR through **(insert theory here)**

Natelytle committed:
Band-aid math error on stable LNs, add tail deviation multiplier

peppy commented:
Is it possible to adjust the test `.osu` file to something that can be understood? ie. it shouldn't be a full beatmap with 8,000 objects, but a limited number of objects where you can parse the file visually and understand the results obtained.

Natelytle committed:
Fix stable LN logic

Eve-ning committed:
Nest Ternary into Math.Log

This removes the need to call Math.Log twice, improving readability.
I did consider doing NoteCount + (isLeg ? 0 : HoldNoteCount), but I think it deproves readability

Eve-ning committed:
Shorten property fetching code

Eve-ning committed:
Fix incorrect function name logPNote

The function actually returns the natural log of the probability of each note.

Eve-ning committed:
Fix incorrect legacyLogPHold function name

Also append that this is only used for legacy holds

Natelytle committed:
Merge master

Eve-ning committed:
Collapse boilerplate code to parse judgements

Eve-ning committed:
Remove resource dependency on test

Eve-ning committed:
Move test case code closer to test callables

Eve-ning committed:
Expose getHitWindows for tests

Eve-ning committed:
Narrow scope of hitWindows fns

Eve-ning committed:
Narrow scope of computeEstimatedUr

We narrow it because maniaAttributes is a large dictionary.
If we can narrow down to the primitives that drives computeEstimatedUr, it makes it easier end-users to supply minimal parameters necessary for it to operate.

Eve-ning committed:
Update changed values

Eve-ning committed:
Add remark on null & exception throwable

Eve-ning committed:
Improve testing

This huge change includes
1) Adding summaries on the purpose of each change
2) Improving the test surface by include more parameters
3) Reducing "Magic Constant" usages

Eve-ning committed:
Add comment justifying args on TestEdge

Eve-ning committed:
Delete ur-estimation-test.osu

Eve-ning committed:
Remove unresolvable cref

Eve-ning committed:
Add test for More Max Less UR

Natelytle committed:
Switch to using deviation on notes and heads for multiplier scaling

smoogipoo committed:
Remove IsConvert difficulty attribute

smoogipoo committed:
Store difficulty attribute values

Eve-ning committed:
Update ManiaUnstableRateEstimationTest.cs

Natelytle committed:
Merge pull request #1 from Eve-ning/maniastatacc-improve-test

Improves `maniastatacc` branch for Tests

Natelytle committed:
Fix non-existent attribute call

Natelytle commented:
Tests should be more manageable now.

smoogipoo commented:
I've run an SR/PP spreadsheet for these changes, and attached it to the OP. Please have a look over it to make sure that the values are expected :)

Natelytle commented:
Sheet looks fine, I cross checked a few values and theyre all fine

Eve-ning reviewed (approved):
LGTM

Natelytle committed:
Merge remote-tracking branch 'osumaster/master' into maniastatacc

Natelytle committed:
Remove hit windows scaling by rate, adjust tests

Natelytle commented:
Will require reapproval due to https://github.com/ppy/osu/pull/24636

smoogipoo committed:
Merge branch 'master' into maniastatacc

Natelytle committed:
Merge master

Natelytle committed:
Increase readability

Natelytle committed:
Merge branch 'ppy:master' into maniastatacc

smoogipoo committed:
Merge branch 'master' into maniastatacc

smoogipoo commented:
Just to be safe...

!diffcalc
RULESET=mania

bdach commented on osu.Game.Rulesets.Mania/osu.Game.Rulesets.Mania.csproj:
```diff
@@ -15,4 +15,8 @@
     <ProjectReference Include="..\osu.Game\osu.Game.csproj" />
   </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="MathNet.Numerics" Version="5.0.0" />
+  </ItemGroup>
```
How feasible is it to eliminate this mathnet reference in a way similar to https://github.com/ppy/osu/pull/20963?

Yes I keep banging on about that and it's probably inconvenient but it matters if it's going to decrease binary size / dependency blast radius.

Natelytle replied:
Yes, tsunyoku ported the relevant functions to a file. I just have to clean up the implementation

bdach commented on osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceAttributes.cs:
```diff
@@ -12,6 +12,12 @@ public class ManiaPerformanceAttributes : PerformanceAttributes
         public double Difficulty { get; set; }
 
+        [JsonProperty("estimated_ur")]
+        public double? EstimatedUr { get; set; }
```
Same as https://github.com/ppy/osu/pull/20963, would prefer this to be called `EstimatedUnstableRate`

Natelytle replied:
Sure.

bdach commented on osu.Game.Rulesets.Mania/Difficulty/ManiaDifficultyAttributes.cs:
```diff
@@ -34,6 +53,9 @@ public override void FromDatabaseAttributes(IReadOnlyDictionary<int, double> val
 
             StarRating = values[ATTRIB_ID_DIFFICULTY];
             GreatHitWindow = values[ATTRIB_ID_GREAT_HIT_WINDOW];
+            OverallDifficulty = values[ATTRIB_ID_OVERALL_DIFFICULTY];
+            NoteCount = (int)values[ATTRIB_ID_NOTE_COUNT];
+            HoldNoteCount = (int)values[ATTRIB_ID_HOLD_NOTE_COUNT];
```
I was going to ask why these are not just being read from the `IBeatmapOnlineInfo` param here and skipped entirely in `.ToDatabaseAttributes()`, but then realised that wouldn't work because of keymods. I'm not sure how intentional this was but yeah.

This is going to be something that we need to be wary of going forward unless realtime calculation becomes a thing.

Natelytle replied:
This part in particular was added by smoogi to calculate a sheet for the rework

bdach commented:
Deployment considerations:

- 3 added difficulty attributes
  - 3 rows per 13 B
  - we have ~19400 mania-specific beatmaps, for which only 16 mod combinations need to be considered = ~12 MB of storage overhead
  - plus ~111700 converts, which have 16 * 14 available mod combinations (keymods) = ~976 MB of storage overhead

Seemingly nothing further to point out from the infra side.

Natelytle commented:
@bdach Before you start any code structure reviews, what are your passing thoughts on the structure of https://github.com/Natelytle/osu/blob/maniastataccrefactor (specifically Utils/LogProb.cs)? I find it's much better for parsing the mathematics since the log stuff is abstracted away, but I'd like a second opinion.

bdach commented:
> what are your passing thoughts on the structure of https://github.com/Natelytle/osu/blob/maniastataccrefactor (specifically Utils/LogProb.cs)

I'm not sure I'm going to be doing math/methodology correctness reviews on diffcalc anymore myself as I'm not sure that it's productive or even needed at this point. I was more angling to do just general code quality stuff. That said, since you asked directly, I'm not sure what to think of that thing. Some of the operations there have me worried about correctness of function domain and/or numerical stability - specifically I'm talking about stuff like presenting

$$ \log (x+y) = \log \left( x \cdot \left( 1 + \frac{y}{x} \right) \right) = \log x + \log \left( 1 + \frac{y}{x} \right) $$

etc (assuming $y > x$). It seems to be *symbolically* correct but I'm not sure how it's going to behave in the wild as I'm rusty on my analysis. Some other operations in there just seem plain arbitrary (what is `LogProb.Combine()`? does it represent some real operation / have any semantic meaning rather than be an arbitrary expression on logs that happens to be useful in this case?)

Again as per my opening post I'd probably not ding this in review because I'm not sure I care about the meat and potatoes of diffcalc changes at this point, but yeah. If these operations are meant to have some probabilistical interpretation then there should be xmldoc that explains what it is, otherwise it just doesn't seem like a very robust abstraction.

Oh and if that thing is gonna stay then please change the name to something resembling full words, `LogProbability` is fine.

Natelytle commented:
Closing due to complexity and the lack of a real use for this in the system.
